content/learning-paths/mobile-graphics-and-gaming/android_halide/fusion.md

```diff
@@ -0,0 +1,349 @@
+
+The directive thresholded.parallel(y) further enhances efficiency by parallelizing the outer loop across multiple CPU cores, thus accelerating execution.
+
+The original scheduling directive used in our example
+
+```cpp
+blur.compute_at(thresholded, x);
+```
+
+instructs Halide to compute the blur function at each iteration of the innermost loop (over the horizontal coordinate x) of the thresholded function. However, as the reviewer correctly pointed out, scheduling at the innermost loop (x) in Halide often results in behavior similar to the default inline scheduling. Yet, explicitly using compute_at(thresholded, x) can introduce additional overhead, such as repeatedly accessing the same small memory region unnecessarily. This could cause redundant memory access because Halide might write intermediate results to a temporary buffer within this innermost loop.
+
+To demonstrate more efficient fusion explicitly—while avoiding redundant intermediate storage—consider instead scheduling blur at an outer loop (e.g., the vertical coordinate y). Doing so significantly improves performance by:
+* Reducing redundant computations and memory accesses.
+* Explicitly fusing operations at a more cache-efficient granularity.
+
+Here’s how you can update your scheduling:
+```cpp
+Halide::Var x("x"), y("y");
+
+// Existing blur definition.
+Halide::Func blur("blur");
+Halide::Func thresholded("thresholded");
+
+// Thresholding definition remains unchanged.
+thresholded(x, y) = Halide::select(blur(x, y) > 128, 255, 0);
+
+// Improved fusion: compute blur at each iteration of outer loop (y).
+blur.compute_at(thresholded, y);
+
+// Optionally, parallelize across the outer loop for improved multicore utilization.
+thresholded.parallel(y);
+```
+
+In the improved code snippet:
+* blur.compute_at(thresholded, y) explicitly instructs Halide to compute blurred rows just before thresholding each row. This means the intermediate blur results are stored temporarily in a small per-row buffer, significantly reducing redundant computations and repeated memory accesses.
+* By choosing y (the vertical loop) instead of the innermost loop (x), we explicitly balance between complete inlining (which may recompute identical results many times) and storing intermediate results explicitly in memory. This strikes a more effective trade-off, avoiding redundant computations while maintaining good cache locality.
+* The thresholded.parallel(y) directive parallelizes the computation across multiple rows, taking full advantage of multi-core CPUs to further accelerate processing.
```

Comment:
I suggest to remove `parallel` here to focus on fusion.
cf. parallel is not always the available option. In some case, one have to stick to single CPU core.